n3fit/src/n3fit/backends/keras_backend/MetaModel.py

```diff
@@ -325,11 +326,133 @@ def reset_layer_weights_to(self, layer_names, reference_vals):
                 w.assign(v)
 
     def apply_as_layer(self, x):
-        """ Apply the model as a layer """
-        all_input = {**self.tensors_in, **x}
+        """Apply the model as a layer"""
+        all_input = {**self.input_tensors, **x}
         return all_input, super().__call__(all_input)
 
     def get_layer_re(self, regex):
-        """ Get all layers matching the given regular expression """
+        """Get all layers matching the given regular expression"""
         check = lambda x: re.match(regex, x.name)
         return list(filter(check, self.layers))
+
+    def get_replica_weights(self, i_replica):
+        """
+        Get the weights of replica i_replica.
+
+        This assumes that the only weights are in layers called
+        ``NN_{i_replica}`` and ``preprocessing_factor_{i_replica}``
+
+
+        Parameters
+        ----------
+            i_replica: int
+
+        Returns
+        -------
+            dict
+                dictionary with the weights of the replica
+        """
+        NN_weights = [
+            tf.Variable(w, name=w.name) for w in self.get_layer(f"{NN}_{i_replica}").weights
+        ]
+        prepro_weights = [
+            tf.Variable(w, name=w.name)
+            for w in self.get_layer(f"{preprocessing}_{i_replica}").weights
+        ]
+        weights = {NN: NN_weights, preprocessing: prepro_weights}
+
+        return weights
+
+    def set_replica_weights(self, weights, i_replica=0):
+        """
+        Set the weights of replica i_replica.
+
+        This assumes that the only weights are in layers called
+        ``NN_{i_replica}`` and ``preprocessing_factor_{i_replica}``
+
+        Parameters
+        ----------
+            weights: dict
+                dictionary with the weights of the replica
+            i_replica: int
+                the replica number to set, defaulting to 0
+        """
+        self.get_layer(f"{NN}_{i_replica}").set_weights(weights[NN])
+        self.get_layer(f"{preprocessing}_{i_replica}").set_weights(weights[preprocessing])
+
+    def split_replicas(self):
+        """
+        Split the single multi-replica model into a list of separate single replica models,
+        maintaining the current state of the weights.
+
+        Returns
+        -------
+            list
+                list of single replica models
+        """
```

Comment:
So that you can have here
```suggestion
        """
        if self.single_replica_generator is None:
            raise ????Error("Trying to generate single replica models with no generator set")
```

(???? because not sure what kind of error would be best)